backend/src/mockExamQuestion/mockExamQuestion.service.ts

```diff
@@ -0,0 +1,224 @@
+    questionsDistribution: QuestionsDistribution,
+  ): Promise<Try<MockExamQuestion[], 'NO_QUESTION_TYPES_FOUND'>> {
+    const {
+      onePointQuestionsAmount,
+      onePointAdvancedQuestionsAmount,
+      twoPointQuestionsAmount,
+      twoPointAdvancedQuestionsAmount,
+      threePointQuestionsAmount,
+      threePointAdvancedQuestionsAmount,
+    } = questionsDistribution;
+    const elementaryQuestionType = await this.questionsTypeRepository.findOne({
+      where: {
+        scope: ELEMENTARY_SCOPE,
+      },
+    });
+    const advancedQuestionType = await this.questionsTypeRepository.findOne({
+      where: {
+        scope: ADVANCED_SCOPE,
+      },
+    });
+    if (!elementaryQuestionType || !advancedQuestionType) {
+      return getFailure('NO_QUESTION_TYPES_FOUND');
+    }
+    const retrievedQuestions = await Promise.all([
+      this.getRandomQuestionsForType(
+        onePointQuestionsAmount,
+        REQUIRED_AMOUNT_OF_ONE_POINT_QUESTIONS,
+        elementaryQuestionType.id,
+        1,
+      ),
+      this.getRandomQuestionsForType(
+        twoPointQuestionsAmount,
+        REQUIRED_AMOUNT_OF_TWO_POINT_QUESTIONS,
+        elementaryQuestionType.id,
+        2,
+      ),
+      this.getRandomQuestionsForType(
+        threePointQuestionsAmount,
+        REQUIRED_AMOUNT_OF_THREE_POINT_QUESTIONS,
+        elementaryQuestionType.id,
+        3,
+      ),
+      this.getRandomQuestionsForType(
+        onePointAdvancedQuestionsAmount,
+        REQUIRED_AMOUNT_OF_ONE_POINT_ADVANCED_QUESTIONS,
+        advancedQuestionType.id,
+        1,
+      ),
+      this.getRandomQuestionsForType(
+        twoPointAdvancedQuestionsAmount,
+        REQUIRED_AMOUNT_OF_TWO_POINT_ADVANCED_QUESTIONS,
+        advancedQuestionType.id,
+        2,
+      ),
+      this.getRandomQuestionsForType(
+        threePointAdvancedQuestionsAmount,
+        REQUIRED_AMOUNT_OF_THREE_POINT_ADVANCED_QUESTIONS,
+        advancedQuestionType.id,
+        3,
+      ),
+    ]);
+    const questions: MockExamQuestion[] = retrievedQuestions.flat();
+    return getSuccess(questions);
+  }
+
+  async getRandomQuestionsForType(
+    totalAmount: number,
+    requiredAmount: number,
+    typeId: number,
+    points: number,
+  ): Promise<MockExamQuestion[]> {
+    const indexes = this.getRandomIndexes(totalAmount, requiredAmount);
+    const questionsPromises: Promise<MockExamQuestion[]>[] = indexes.map(
+      (index) => {
+        return this.mockExamQuestionRepository.find({
+          where: {
+            points,
+            type: {
+              id: typeId,
+            },
+          },
+          skip: index,
+          take: 1,
+        });
+      },
+    );
+    const questions = await Promise.all(questionsPromises);
+    return questions.flat();
+  }
+
+  getRandomIndexes(
+    totalAmountOfIndexes: number,
+    requiredAmountOfIndexes: number,
+  ): number[] {
+    const indexes: number[] = [];
+    while (indexes.length < requiredAmountOfIndexes) {
```

Comment:
Let's add a sanity check to prevent potential inifite iterations of `while`. Maybe 2 × of `requiredAmountOfIndexes`?